cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -479,22 +479,38 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
             callptr->function_name);
       }
 
-      // TODO: ARROW-16624 Add Suffix support for Substrait
-      const auto* left_keys = callptr->arguments[0].field_ref();
-      const auto* right_keys = callptr->arguments[1].field_ref();
-      if (!left_keys || !right_keys) {
-        return Status::Invalid("Left keys for join cannot be null");
-      }
-
       // Create output schema from left, right relations and join keys
       FieldVector combined_fields = left.output_schema->fields();
       const FieldVector& right_fields = right.output_schema->fields();
       combined_fields.insert(combined_fields.end(), right_fields.begin(),
                              right_fields.end());
       std::shared_ptr<Schema> join_schema = schema(std::move(combined_fields));
 
+      // adjust the join_keys according to Substrait definition where
+      // the join fields are defined by considering the `join_schema` which
+      // is the combination of the left and right relation schema.
+
+      // TODO: ARROW-16624 Add Suffix support for Substrait
+      const auto* left_keys = callptr->arguments[0].field_ref();
+      const auto* right_keys = callptr->arguments[1].field_ref();
+      // Validating JoinKeys
+      if (!left_keys || !right_keys) {
+        return Status::Invalid("Left keys for join cannot be null");
+      }
+      // TODO: verify this validation
+      if (!left_keys->IsFieldPath() || !right_keys->IsFieldPath()) {
+        return Status::Invalid("Join keys must be FieldPaths");
+      }
+      int num_left_fields = left.output_schema->num_fields();
+      const auto* right_field_path = right_keys->field_path();
+      std::vector<int> adjusted_field_indices;
+      adjusted_field_indices.reserve(join_schema->num_fields() - num_left_fields);
+      for (size_t idx = 0; idx < right_field_path->indices().size(); idx++) {
+        adjusted_field_indices.push_back((*right_field_path)[idx] - num_left_fields);
+      }
+      FieldPath adjusted_right_keys(adjusted_field_indices);
```

Comment:
```suggestion
      std::vector<int> adjusted_field_indices(right_field_path->indices());
      adjusted_field_indices[0] -= num_left_fields;
      FieldPath adjusted_right_keys(adjusted_field_indices);
```

We only need to modify the top-level index.  The other indices, if any, refer to struct fields and I'm pretty sure they are relative.

For example, if the fifth column (in the combined schema) is `struct<int32,int32,struct<int32>>` and we want the innermost `int32` as the key and there are 2 left fields then the field path will be `[4, 2, 0]`.  We want to change that to `[2, 2, 0]` and not `[2, 0, -2]`.